src/sentry/middleware/integrations/integration_proxy.py

```diff
@@ -0,0 +1,141 @@
+if this is modified
+"""
+PROXY_OI_HEADER = "X-Sentry-Organization-Integration"
+PROXY_SIGNATURE_HEADER = "X-Sentry-Subnet-Signature"
+PROXY_TIMESTAMP_HEADER = "X-Sentry-Subnet-Timestamp"
+
+
+class IntegrationProxyMiddleware:
+    def __init__(self, get_response):
+        self.get_response = get_response
+
+    @property
+    def full_url(self) -> str:
+        return f"{self.client.base_url}/{self.proxy_path}"
+
+    def _validate_sender(self, request: Request) -> bool:
+        """
+        Returns True if the sender is deemed sufficiently trustworthy.
+        """
+        log_extra = {"path": request.path, "host": request.headers.get("Host")}
+        timestamp = request.headers.get(PROXY_TIMESTAMP_HEADER)
+        signature = request.headers.get(PROXY_SIGNATURE_HEADER)
+        if timestamp is None or signature is None:
+            logger.error("invalid_sender_headers", extra=log_extra)
+            return False
+
+        is_valid = verify_subnet_signature(
+            timestamp=timestamp, request_body=request.body, provided_signature=signature
+        )
+        if not is_valid:
+            logger.error("invalid_sender_signature", extra=log_extra)
+
+        return is_valid
+
+    def _validate_request(self, request: Request) -> bool:
+        """
+        Returns True if a client could be generated from the request
+        """
+        from sentry.shared_integrations.client.proxy import IntegrationProxyClient
+
+        self.proxy_path = request.path[len(PROXY_ADDRESS) :]
+        self.headers = clean_proxy_headers(request.headers)
+
+        log_extra = {"path": self.proxy_path}
+
+        # Get the organization integration
+        org_integration_id = self.headers.pop(f"{PROXY_OI_HEADER}", None)
+        if org_integration_id is None:
+            logger.info("missing_org_integration", extra=log_extra)
+            return False
+        log_extra["org_integration_id"] = org_integration_id
+
+        self.org_integration = OrganizationIntegration.objects.filter(
+            id=org_integration_id,
+            status=ObjectStatus.ACTIVE,
+        ).first()
+        if self.org_integration is None:
+            logger.info("invalid_org_integration", extra=log_extra)
+            return False
+        log_extra["integration_id"] = self.org_integration.integration_id
+
+        # Get the integration
+        self.integration = Integration.objects.filter(
+            status=ObjectStatus.ACTIVE,
+            id=self.org_integration.integration_id,
+        ).first()
+        if self.integration is None:
+            logger.info("invalid_integration", extra=log_extra)
+            return False
+
+        # Get the integration client
+        installation = self.integration.get_installation(
+            organization_id=self.org_integration.organization_id
+        )
+        self.client: IntegrationProxyClient = installation.get_client()
+        client_type = type(self.client)
+        log_extra["client_type"] = client_type
```

Comment:
In this case it's the class object since we use `issubclass()` to ensure it derives from the proxy client. That way we don't accidentally proxy requests for integrations that haven't been intentionally set up for use in hybrid cloud.